### PR TITLE
Fix "Context has expired" error in dictionaries

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix the "Context has expired" error in dictionaries when using subqueries.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit d25961f7f95bdfd8b26a580088c13654be64de7f with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53342/d25961f7f95bdfd8b26a580088c13654be64de7f/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Functions/FunctionsExternalDictionaries.h [hunk 1]
```diff
@@ -65,7 +65,7 @@ namespace ErrorCodes
 class FunctionDictHelper : WithContext
 {
 public:
-    explicit FunctionDictHelper(ContextPtr context_) : WithContext(context_) {}
+    explicit FunctionDictHelper(ContextPtr context_) : WithContext(context_->getGlobalContext()) {}
```
Comment (al13n321) [inline src/Functions/FunctionsExternalDictionaries.h:68]: There's current_context->checkAccess() on line 77 below. Will it work correctly with global context, or does it take the current user's identity from the non-global part of context?

Comment (alexey-milovidov) [reply src/Functions/FunctionsExternalDictionaries.h:68]: Most likely not :(
Let's add a test case.

PS. There is also a session context.

**Review by al13n321 — APPROVED** [review]
Seems reasonable (I don't know anything about any of this code though). I guess it'll still fail if the session gets disconnected at just the wrong moment, but that's unlikely.

Alternatively, it could retain a shared_ptr to `context->getAccess()`, maybe.

**Comment by alexey-milovidov** [comment]
Not every case is ok: https://s3.amazonaws.com/clickhouse-test-reports/53342/96018399808b133f5adc6b477203d5613b2f625e/fast_test.html

Sometimes the functions run without a session.

**Comment by alexey-milovidov** [comment]
It's strange, but it succeeded.
Need help from @tavplubix

**Pull request merged**